README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# spiR

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/warint/spiR.svg?branch=master)](https://travis-ci.org/warint/spiR)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/warint/spiR?branch=master&svg=true)](https://ci.appveyor.com/project/warint/spiR)
[![CRAN
status](https://www.r-pkg.org/badges/version/spiR)](https://CRAN.R-project.org/package=spiR)
[![](https://cranlogs.r-pkg.org/badges/grand-total/spiR?color=blue)](https://cran.r-project.org/package=spiR)
<!-- badges: end -->

In 2015, The 17 United Nations’ Sustainable Development Goals were
adopted. ‘spiR’ is a wrapper of several open datasets published by the
Social Progress Imperative (<https://www.socialprogress.org/>),
including the Social Progress Index (a synthetic measure of human
development across the world). ‘spiR’‘s goal is to provide data to help
policymakers and researchers prioritize actions that accelerate social
progress across the world in the context of the Sustainable Development
Goals. The Social Progress Index proposes a new perspective on social
challenges and needed efforts to accelerate social progress in line with
the Sustainable Development Goals. In this context, the goal of ’spiR’
is to allow an easy connection with R to the Social Progress Index in
order to benefit from the “power of crowds.” ‘spiR’ is an R wrapper to
easily access the Social Progress Index datasets.

At the Social Progress Imperative, they define “social progress as the
capacity of a society to meet the basic human needs of its citizens,
establish the building blocks that allow citizens and communities to
enhance and sustain the quality of their lives, and create the
conditions for all individuals to reach their full potential. Improving
quality of life is a complex task and past efforts to measure progress
simply haven’t created a sufficiently nuanced picture of what a
successful society looks like. That’s why we created the Social Progress
Index. Rather than emphasizing traditional measurements of success like
income and investment, we measure 51 social and environmental indicators
to create a clearer picture of what life is really like for everyday
people. The index doesn’t measure people’s happiness or life
satisfaction, focusing instead on actual life outcomes in areas from
shelter and nutrition to rights and education. This exclusive focus on
measurable outcomes makes the index a useful policy tool that tracks
changes in society over time.”

The index measures the quality of life for 98% of the world’s
population. In its current version, the R client provides access to
global data. In further versions, this R client will include different
geographical levels: states, regions, cities and sometimes communities.

Three overarching dimensions are (1) Basic Human Needs, (2) Foundations
of Wellbeing, and (3) Opportunity. Within each dimension, there are four
components that further divide the indicators into thematic categories.
The index consists in 51 social and environmental variables, covering
the years 2014 to 2019.

To see a definition of the indicators, please visit this
[page](https://www.socialprogress.org/index/global/definitions/#BHN/0/0/).

To read the methodology summary of the Social Progress Index
[\[here\]](https://legacy.socialprogress.org/assets/downloads/resources/2020/2020-Social-Progress-Index-Methodology.pdf)

See below to have access through ‘spiR’ to the whole list of indicators.

## Overview of the available data

Users can recreate in R impactful dashboards and visualizations as the
ones found on [www.socialprogress.org](https://www.socialprogress.org/):
[click
here](https://www.socialprogress.org/?tab=2&code=NOR&embedded=true/)

## Installation

The released version of spiR package is accessible through CRAN and
devtools.

``` r
install.packages("spiR")

install.packages("devtools")
devtools::install_github('warint/spiR')
```

## How-To

### Step 1: Getting the country’s ISO code

A user needs to enter the ISO code of a country. To have access to this
code, the following function provides this information.

``` r
spir_country() # A list of all countries will be produced

spir_country(country = "Canada") # The ISO code for Canada will be produced

spir_country("Canada") # The ISO code for Canada will be produced
```

### Step 2: Getting the indicator’s code

A user needs to enter the code of the desired indicator. To do so, the
following function provides access to all the indicators of interest.

``` r
spir_indicator() # A list of all indicators will be produced

spir_indicator(indicators = "mortality") # A list with all the variables including "mortlity" will be produced

spir_indicator("mortality") # A list with all the variables including "mortlity" will be produced
```

### Step 3: Getting the data

Once the user knows the ISO code and the indicator’s code, s.he can
collect the data in a very easy way through this function:

``` r
spir_data(country = c("USA", "FRA"), year = c("2018", "2019"), indicators = "SPI") # It generates a data frame of the overall SPI indicator for the USA and France for the years 2018 and 2019

spir_data(country = c("USA", "FRA"), years = "2018", ) # It generates a data frame of all the indicators for the USA and France for the year 2018

spir_data("USA", "2019", c("SPI", "FOW")) # It generates a data frame of the SPI and FOW indicators for the USA in 2019

spir_data(, "2018", ) # It generates a data frame for all the countries and all the indicators for 2018

spir_data("USA", "2017", ) # It generates a data frame of all the indicators for the USA in 2017

spir_data("USA", , ) # It generates a data frame of all the indicators for all the years for the USA

spir_data(, , ) # It generates a data frame of the complete dataset

spir_data() # It generates a data frame of the complete dataset
```

### Optional Step : Create Visuals

spir\_visual fonction allows to create two types of visuals from the SPI
data : barchart and linechart.

The first is a barchart of the top 9 countries for the chosen indicator
and year.

<img src="man/figures/README-unnamed-chunk-5-1.png" width="100%" />

The second is a linechart showing the evolution of the most represented
countries over time for the chosen indicator.

<img src="man/figures/README-unnamed-chunk-6-1.png" width="100%" />

### Cite ‘spiR’

If you use the spiR package for research, education, policymaking or
other related purposes, please cite the package as well as the Social
Progress Imperative. Consider making them a donation to help them pursue
the work. Business, social enterprises and investors interested in using
the Social Progress Index should contact their offices.

To cite the ‘spiR’ package in your work:

Warin, T. (2019). “spiR: Client for the Social Progress Index. v0.2.1.”

``` r
@article{Warin2019,
author = "Thierry Warin",
title = "{spiR: An R Package for the Social Progress Index}",
year = "2019",
month = "12",
url = "https://figshare.com/articles/spiR_Client_for_Social_Progress_Index_Data/11421573/",
doi = "10.6084/m9.figshare.11421573.v2"
}
```

### Acknowledgements

The author would like to thank the Center for Interuniversity Research
and Analysis of Organizations (CIRANO, Montreal) for its support, as
well as Thibault Senegas, Marine Leroi and Martin Paquette. The usual
caveats apply.


NEWS.md
# spiR 0.2.1 (2021-01-05)

* In this version, we have changed the name of the functions by withdrawing the sqs mentions. The old functions will remain available until the next update (deprecated).

* We have added a new function, spir_visual, allowing people to create visuals from the SPI data. 


R/spiData.R
# Function 1: Data collection

SPI_data <- gsheet::gsheet2tbl("https://docs.google.com/spreadsheets/d/1_nQ9mQU_4J0KDRc4_TMzTsJHMYBqLwwnPaMC5BVhkGc/edit#gid=0")

data_long <- reshape2::melt(SPI_data,
  # ID variables - all the variables to keep but not split apart on
  id.vars = c("countryName", "code", "year"),
  # The source columns
  measure.vars = colnames(SPI_data)[6:ncol(SPI_data)],
  # Name of the destination column that will identify the original
  # column that the measurement came from
  variable.name = "indicator",
  value.name = "value"
)

# Creating the default values for the function query

data_long_country <- unique(data_long[, 2])
data_long_year <- unique(data_long[, 3])
data_long_indicator <- unique(data_long[, 4])

sqs_spi_data <- function(country = data_long_country, years = data_long_year, indicators = data_long_indicator) {
  out <- dplyr::filter(data_long, code %in% country & year %in% years & indicator %in% indicators)

  return(out)
}

# Function 2: Indicators' symbols reconciliation

sqs_spi_symbol <- function(indicators) {
  SPI_Indicators <- gsheet::gsheet2tbl("https://docs.google.com/spreadsheets/d/1_nQ9mQU_4J0KDRc4_TMzTsJHMYBqLwwnPaMC5BVhkGc/edit#gid=400714513")
  if (missing(indicators)) {
    SPI_Indicators
  } else {
    SPI_Indicators[grep(indicators, SPI_Indicators$indicator_name), ]
  }
}

# Function 3: Countries' code reconciliation

sqs_spi_country <- function(country) {
  SPI_data <- gsheet::gsheet2tbl("https://docs.google.com/spreadsheets/d/1_nQ9mQU_4J0KDRc4_TMzTsJHMYBqLwwnPaMC5BVhkGc/edit#gid=0")
  SPI_country <- unique(SPI_data[, 1:2])
  if (missing(country)) {
    SPI_country
  } else {
    SPI_country[grep(country, SPI_country$countryName), ]
  }
}


R/spiSymbol.R

#' sqs_spi_symbol
#'
#' @description This function allows you to find and search the right indicator code from the Social Progress Index you want to use.
#'
#' If no argument is filed, all indicators will be displayed.
#'
#' @param INDICATOR A social indicator from the Social Progress Index.
#'
#' @return Indicator code from the Social Progress Index.
#' @export
#' @import gsheet
#'
#'
#'@seealso \code{\link{sqs_spi_country}} for the SPI's country code and \code{\link{sqs_spi_data}} to collect data when you have both the indicator and the country code.
#'
#' @examples
#' sqs_spi_symbol("electricity")
#' AllIndicators <- sqs_spi_symbol()
sqs_spi_symbol <- function(INDICATOR){

  SPI_Indicators <- gsheet2tbl("https://docs.google.com/spreadsheets/d/1_nQ9mQU_4J0KDRc4_TMzTsJHMYBqLwwnPaMC5BVhkGc/edit#gid=400714513")

  if (missing(INDICATOR)) {

    SPI_Indicators

  } else {

    SPI_Indicators[grep(INDICATOR,SPI_Indicators$indicator_name),]
  }

}


R/spiCountry.R
#' sqs_spi_country
#'
#'@description This function allows you to find and search the right country code associated with the Social Progress Index's Data.
#'
#'If no argument is filed, all countries will be displayed.
#'
#' @param COUNTRY The name of the country starting with a capital letter.
#'
#' @return A country code
#' @export
#' @import gsheet
#'
#' @seealso \code{\link{sqs_spi_symbol}} for the SPI's indicator symbol and \code{\link{sqs_spi_data}} to collect data when you have both the indicator and the country code.
#'
#' @examples
#' sqs_spi_country("Canada")
#' AllCountries <- sqs_spi_country()
sqs_spi_country <- function(COUNTRY){

  SPI_data <- gsheet2tbl("https://docs.google.com/spreadsheets/d/1_nQ9mQU_4J0KDRc4_TMzTsJHMYBqLwwnPaMC5BVhkGc/edit#gid=0")
  SPI_country <- unique(SPI_data[,1:2])

  if (missing(COUNTRY)) {

    SPI_country

  } else {

    SPI_country[grep(COUNTRY,SPI_country$country),]
  }
}


R/spir_newfunctions.R
library(dplyr)
library(ggplot2)
library(ggsci)

# Loading data
url <- paste0("https://warin.ca/datalake/spiR/SPI_data.csv")
path <- file.path(tempdir(), "temp.csv")
curl::curl_download(url, path)
csv_file <- file.path(paste0(tempdir(), "/temp.csv"))
SPI_data <- readr::read_csv(csv_file)

# Loading indicators
url <- paste0("https://warin.ca/datalake/spiR/SPI_indicator.csv")
path <- file.path(tempdir(), "temp.csv")
curl::curl_download(url, path)
csv_file <- file.path(paste0(tempdir(), "/temp.csv"))
SPI_indicator <- readr::read_csv(csv_file)

# Creating the default values for the function query
# IF an entry is missing, all the observations of this variable will be displayed

spi_country <- base::unique(SPI_data$country_code)
spi_year <- base::unique(SPI_data$year)
spi_indicator <- base::unique(SPI_data$indicator_code)


# Function 1: Data collection

spir_data <- function(country = spi_country,
                     years = spi_year,
                     indicators = spi_indicator) {
  country_code <- year <- indicator_code <- NULL
  out <- dplyr::filter(SPI_data,
                       country_code %in% country,
                       year %in% years,
                       indicator_code %in% indicators)
  return(out)
}


# Function 2: Indicators' symbols query

spir_indicator <- function(indicators) {
  spi_indicators_natural_language <- SPI_indicator
  if (missing(indicators)) {
    spi_indicators_natural_language
  } else {
    spi_indicators_natural_language[grep(indicators, spi_indicators_natural_language$indicator_name, ignore.case = TRUE), ]
  }
}


# Function 3: Countries' code reconciliation

spir_country <- function(country) {
  spi_countries_natural_language <- unique(SPI_data[, 2:3])
  if (missing(country)) {
    spi_countries_natural_language
  } else {
    spi_countries_natural_language[grep(country, spi_countries_natural_language$country_name, ignore.case = TRUE), ]
  }
}



# Function 4: Visualization

spir_visual <- function(chart = "bar_1", indicator = "SPI", years = max(SPI_data$year), title = TRUE){
  
  if(years > 2020 | years < 2011){
    stop("no available data for the requested year")
  }
  
  if(chart == "bar_1"){
    barchart1 <- SPI_data
    barchart1 <- filter(barchart1, year == years)
    barchart1 <- filter(barchart1, indicator_code == indicator)
    barchart1 <- arrange(barchart1, desc(value))
    countries <- barchart1[1:9, 3]
    countries <- countries$country_code
    barchart1 <- dplyr::filter(barchart1, country_code %in% countries)
    barchart1$country_name <- factor(barchart1$country_name, levels = unique(barchart1$country_name)[order(barchart1$value)])
    barchart1$value <- format(round(barchart1$value, 2), nsmall = 2)
     if(title == TRUE){
      ggplot(data = barchart1, aes(x = country_name, y = value, fill = country_name)) +
        geom_col() +
        geom_text(aes(label=value), vjust=0.2, hjust = 1.1, colour = "white", size = 3.2, fontface = "bold") + 
        ylab("")  +
        xlab("") +
        ggtitle(paste(unique(barchart1$indicator_name), "in", years)) +
        theme_minimal() +
        guides(fill=FALSE) +
        scale_fill_uchicago() +
        theme(legend.position="none", plot.title = element_text(size=12))  + 
        labs(fill = "Countries", caption="Source: Warin (2020) & Social Progress Index.") + 
        coord_flip() +
         theme(axis.title.x=element_blank(),
               axis.ticks.x=element_blank(),
               axis.text.x=element_blank())
    } else {
      ggplot(data = barchart1, aes(x = country_name, y = value, fill = country_name)) +
        geom_col() +
        geom_text(aes(label=value), vjust=0.2, hjust = 1.1, colour = "white", size = 3.2, fontface = "bold") + 
        ylab("")  +
        xlab("") +
        ggtitle("") +
        theme_minimal() +
        guides(fill=FALSE) +
        scale_fill_uchicago() +
        theme(legend.position="none", plot.title = element_text(size=12))  + 
        labs(fill = "Countries", caption="Source: Warin (2020) & Social Progress Index.") + 
        coord_flip() +
        theme(axis.title.x=element_blank(),
              axis.text.x=element_blank(),
              axis.ticks.x=element_blank())
    }
  } else if(chart == "line_1"){
    linechart1 <- SPI_data
    linechart1 <- filter(linechart1, indicator_code == indicator)
    linechart1 <- arrange(linechart1, desc(year), desc(value))
    countries <- linechart1[1:6, 3]
    countries <- countries$country_code
    linechart1 <- dplyr::filter(linechart1, country_code %in% countries)
    linechart1$year <- lubridate::ymd(linechart1$year, truncated = 2L)
    if(title == TRUE){
      ggplot(data = linechart1, aes(x = year, y = value, color = country_name, shape = country_name)) +
        geom_line() + 
        geom_point(size = 2, stroke = 1) + 
        ylab("")  +
        xlab("") +
        ggtitle(paste0(unique(linechart1$indicator_name), ": \nEvolution of the ", max(SPI_data$year), " most represented countries")) +
        theme_minimal() +
        guides(fill=FALSE) +
        scale_color_uchicago() +
        theme(legend.position="right", plot.title = element_text(size=12))  + 
        labs(shape = "Countries", color = "Countries", caption="Source: Warin (2020) & Social Progress Index.")
    } else {
      ggplot(data = linechart1, aes(x = year, y = value, color = country_name, shape = country_name)) +
        geom_line() + 
        ylab("")  +
        xlab("") +
        ggtitle("") +
        theme_minimal() +
        guides(fill=FALSE) +
        geom_point(size = 2, stroke = 1) + 
        scale_color_uchicago() +
        theme(legend.position="right", plot.title = element_text(size=12))  + 
        labs(shape = "Countries", color = "Countries", caption="Source: Warin (2020) & Social Progress Index.")
    }
  } else{
    stop("invalid arguments")
  }
}



R/spiR.R


# Loading data
url <- paste0("https://warin.ca/datalake/spiR/SPI_data.csv")
path <- file.path(tempdir(), "temp.csv")
curl::curl_download(url, path)
csv_file <- file.path(paste0(tempdir(), "/temp.csv"))
SPI_data <- readr::read_csv(csv_file)


# Loading indicators
url <- paste0("https://warin.ca/datalake/spiR/SPI_indicator.csv")
path <- file.path(tempdir(), "temp.csv")
curl::curl_download(url, path)
csv_file <- file.path(paste0(tempdir(), "/temp.csv"))
SPI_indicator <- readr::read_csv(csv_file)

# Creating the default values for the function query
# IF an entry is missing, all the observations of this variable will be displayed

spi_country <- base::unique(SPI_data$country_code)
spi_year <- base::unique(SPI_data$year)
spi_indicator <- base::unique(SPI_data$indicator_code)


# Function 1: Data collection

#' spir_data
#'
#' @description This function allows you to find and display the Social Progress Index data according to the selected parameters.
#' If no arguments are filled, all data will be displayed.
#'
#' @param country Countries' ISO code.
#' @param years Years for which you want data.
#' @param indicators Indicators from the Social Progress Index.
#' @import gsheet
#' @import dplyr
#' @import reshape2
#' @return Data for the country, indicator and year requested.
#' @export
#'
#' @seealso \code{\link{spir_indicator}} for the SPI's indicator symbol and \code{\link{spir_country}} for the country's ISO code.
#'
#' @examples
#' myData<- spir_data(country = c("USA", "FRA"), years = "2018", )
#' myData<- spir_data(country = c("USA", "FRA"), year = c("2018", "2019"), indicators = "SPI")
#' myData<- spir_data("USA", "2019", c("SPI", "FOW"))
#' myData<- spir_data(, "2018", )
#' myData<- spir_data("USA", "2017", )
#' myData<- spir_data("USA", , )
#' myData<- spir_data(, , )
#' myData<- spir_data()


spir_data <- function(country = spi_country,
                     years = spi_year,
                     indicators = spi_indicator) {
  country_code <- year <- indicator_code <- NULL
  out <- dplyr::filter(SPI_data,
                       country_code %in% country,
                       year %in% years,
                       indicator_code %in% indicators)
  return(out)
}


# Function 1 (deprecated): Data collection

#' sqs_spi_data (deprecated)
#'
#' @description This function allows you to find and display the Social Progress Index data according to the selected parameters.
#' If no arguments are filled, all data will be displayed. This function will be removed in the next version and replaced by the simpler function 'spir_data()'
#'
#' @param country Countries' ISO code.
#' @param years Years for which you want data.
#' @param indicators Indicators from the Social Progress Index.
#' @import gsheet
#' @import dplyr
#' @import reshape2
#'
#' @name spiR_data-deprecated
#'
#' @return Data for the country, indicator and year requested.
#' @export
#'
#' @seealso \code{\link{spir_indicator}} for the SPI's indicator symbol and \code{\link{spir_country}} for the country's ISO code.
#'
#' @examples
#' myData<- sqs_spi_data(country = c("USA", "FRA"), years = "2018", )
#' myData<- sqs_spi_data(country = c("USA", "FRA"), year = c("2018", "2019"), indicators = "SPI")
#' myData<- sqs_spi_data("USA", "2019", c("SPI", "FOW"))
#' myData<- sqs_spi_data(, "2018", )
#' myData<- sqs_spi_data("USA", "2017", )
#' myData<- sqs_spi_data("USA", , )
#' myData<- sqs_spi_data(, , )
#' myData<- sqs_spi_data()


sqs_spi_data <- function(country = spi_country,
                      years = spi_year,
                      indicators = spi_indicator) {
  .Deprecated(msg = "'sqs_spi_data()' will be removed in the next version and replaced by the simpler function 'spir_data()'")

  # In the next version, we will uncomment the next line:
  # .Defunct(msg = "'sqs_spi_data()' has been removed from this package and replaced with 'spir_data()'")

  country_code <- year <- indicator_code <- NULL
  out <- dplyr::filter(SPI_data,
                       country_code %in% country,
                       year %in% years,
                       indicator_code %in% indicators)
  return(out)
}



# Function 2: Indicators' symbols query
# If the user does not know the code of an indicator, s.he has access to the answer in natural language through this query

#' spir_indicator
#'
#' @description This function allows you to find and search the right indicator code from the Social Progress Index you want to use.
#' If no argument is filed, all indicators will be displayed.
#' @param indicators An indicator from the Social Progress Index.
#'
#' @return Indicator code from the Social Progress Index.
#' @export
#' @seealso \code{\link{spir_country}} for the SPI's country code and \code{\link{spir_data}} to collect data when you have both indicator and country code.
#'
#' @examples
#'myIndicator <- spir_indicator()
#'myIndicator <- spir_indicator(indicators = "mortality")
#'myIndicator <- spir_indicator("mortality")



spir_indicator <- function(indicators) {
  spi_indicators_natural_language <- SPI_indicator
  if (missing(indicators)) {
    spi_indicators_natural_language
  } else {
    spi_indicators_natural_language[grep(indicators, spi_indicators_natural_language$indicator_name, ignore.case = TRUE), ]
  }
}





# Function 2 (deprecated): Indicators' symbols query
# If the user does not know the code of an indicator, s.he has access to the answer in natural language through this query

#' sqs_spi_indicator(deprecated)
#'
#' @description This function allows you to find and search the right indicator code from the Social Progress Index you want to use.
#' If no argument is filed, all indicators will be displayed. This function will be removed in the next version and replaced by the simpler function 'spir_indicator()'
#'
#' @param indicators An indicator from the Social Progress Index.
#'
#' @return Indicator code from the Social Progress Index.
#' @export
#' @seealso \code{\link{spir_country}} for the SPI's country code and \code{\link{spir_data}} to collect data when you have both indicator and country code.
#'
#' @name spiR_indicator-deprecated
#'
#' @examples
#'myIndicator <- sqs_spi_indicator()
#'myIndicator <- sqs_spi_indicator(indicators = "mortality")
#'myIndicator <- sqs_spi_indicator("mortality")

sqs_spi_indicator <- function(indicators) {
  .Deprecated(msg = "'sqs_spi_indicator()' will be removed in the next version and replaced by the simpler function 'spir_indicator()'")

  # In the next version, we will uncomment the next line:
  # .Defunct(msg = "'sqs_spi_indicator()' has been removed from this package and replaced with 'spir_indicator()'")

  spi_indicators_natural_language <- SPI_indicator
  if (missing(indicators)) {
    spi_indicators_natural_language
  } else {
    spi_indicators_natural_language[grep(indicators, spi_indicators_natural_language$indicator_name, ignore.case = TRUE), ]
  }
}



# Function 3: Countries' code reconciliation
# If the user does not know the ISO code of a country, s.he has access to the answer in natural language through this query

#' spir_country
#'
#' @description This function allows you to find and search the right country code associated with the Social Progress Index's Data.
#' If no argument is filed, all indicators will be displayed.
#'
#' @param country The name of the country.
#'
#' @return Country's ISO code.
#'
#' @export
#' @seealso \code{\link{spir_indicator}} for the SPI's indicators and \code{\link{spir_data}} to collect data when you have both indicator and country code.
#' @examples
#'mycountry <- spir_country()
#'mycountry <- spir_country(country = "Canada")
#'mycountry <- spir_country("Canada")
#'

spir_country <- function(country) {
  spi_countries_natural_language <- unique(SPI_data[, 2:3])
  if (missing(country)) {
    spi_countries_natural_language
  } else {
    spi_countries_natural_language[grep(country, spi_countries_natural_language$country_name, ignore.case = TRUE), ]
  }
}

# Function 3 (deprecated): Countries' code reconciliation
# If the user does not know the ISO code of a country, s.he has access to the answer in natural language through this query

#' sqs_spi_country (deprecated)
#'
#' @description This function allows you to find and search the right country code associated with the Social Progress Index's Data.
#' If no argument is filed, all indicators will be displayed. This function will be removed in the next version and replaced by the simpler function 'spir_country()'
#'
#' @param country The name of the country.
#' @name spiR_country-deprecated
#' @return Country's ISO code.
#' @export
#'
#' @seealso \code{\link{spir_indicator}} for the SPI's indicators and \code{\link{spir_data}} to collect data when you have both indicator and country code.
#' @examples
#'mycountry <- sqs_spi_country()
#'mycountry <- sqs_spi_country(country = "Canada")
#'mycountry <- sqs_spi_country("Canada")
#'

sqs_spi_country <- function(country) {
  .Deprecated(msg = "'sqs_spi_country()' will be removed in the next version and replaced by the simpler function 'spir_country()'")

  # In the next version, we will uncomment the next line:
  # .Defunct(msg = "'sqs_spi_country()' has been removed from this package and replaced with 'spir_country()'")
  spi_countries_natural_language <- unique(SPI_data[, 2:3])
  if (missing(country)) {
    spi_countries_natural_language
  } else {
    spi_countries_natural_language[grep(country, spi_countries_natural_language$country_name, ignore.case = TRUE), ]
  }
}

# Function 4: Visualization

#' spir_visual
#'
#' @description This function allows you to create 2 types of visuals: bar and line charts.
#' @param chart Type of charts.
#' @param indicator An indicator from the Social Progress Index.
#' @param years A chosen year
#' @param title Chart title, set by default to TRUE
#'
#' @return Chosen Graph
#'
#' @import dplyr
#' @import ggplot2
#' @import ggsci
#' @import lubridate
#'
#' @export
#'
#' @examples
#'spir_visual(chart = "bar_1", indicator = "SPI", years = "2020")

spir_visual <- function(chart = "bar_1", indicator = "SPI", years = max(SPI_data$year), title = TRUE){

  year <- value <- indicator_code <- country_name <- country_code <- NULL

  if(years > 2020 | years < 2011){
    stop("no available data for the requested year")
  }

  if(chart == "bar_1"){
    barchart1 <- SPI_data
    barchart1 <- dplyr::filter(barchart1, year == years)
    barchart1 <- dplyr::filter(barchart1, indicator_code == indicator)
    barchart1 <- dplyr::arrange(barchart1, desc(value))
    countries <- barchart1[1:9, 3]
    countries <- countries$country_code
    barchart1 <- dplyr::filter(barchart1, country_code %in% countries)
    barchart1$country_name <- factor(barchart1$country_name, levels = unique(barchart1$country_name)[order(barchart1$value)])
    barchart1$value <- format(round(barchart1$value, 2), nsmall = 2)
     if(title == TRUE){
       ggplot2::ggplot(data = barchart1, ggplot2::aes(x = country_name, y = value, fill = country_name)) +
         ggplot2::geom_col() +
         ggplot2::geom_text(ggplot2::aes(label=value), vjust=0.2, hjust = 1.1, colour = "white", size = 3.2, fontface = "bold") +
         ggplot2::ylab("")  +
         ggplot2::xlab("") +
         ggplot2::ggtitle(paste(unique(barchart1$indicator_name), "in", years)) +
         ggplot2::theme_minimal() +
         ggplot2::guides(fill=FALSE) +
         ggsci::scale_fill_uchicago() +
         ggplot2::theme(legend.position="none", plot.title = ggplot2::element_text(size=12))  +
         ggplot2::labs(fill = "Countries", caption="Source: Warin (2020) & Social Progress Index.") +
         ggplot2::coord_flip() +
         ggplot2::theme(axis.title.x=ggplot2::element_blank(),
               axis.ticks.x=ggplot2::element_blank(),
               axis.text.x=ggplot2::element_blank())
    } else {
      ggplot2::ggplot(data = barchart1, ggplot2::aes(x = country_name, y = value, fill = country_name)) +
        ggplot2::geom_col() +
        ggplot2::geom_text(ggplot2::aes(label=value), vjust=0.2, hjust = 1.1, colour = "white", size = 3.2, fontface = "bold") +
        ggplot2::ylab("")  +
        ggplot2::xlab("") +
        ggtitle("") +
        ggplot2::theme_minimal() +
        ggplot2::guides(fill=FALSE) +
        ggsci::scale_fill_uchicago() +
        ggplot2:: theme(legend.position="none", plot.title = ggplot2::element_text(size=12))  +
        ggplot2::labs(fill = "Countries", caption="Source: Warin (2020) & Social Progress Index.") +
        ggplot2::coord_flip() +
        ggplot2::theme(axis.title.x=ggplot2::element_blank(),
              axis.text.x=ggplot2::element_blank(),
              axis.ticks.x=ggplot2::element_blank())
    }
  } else if(chart == "line_1"){
    linechart1 <- SPI_data
    linechart1 <- dplyr::filter(linechart1, indicator_code == indicator)
    linechart1 <- dplyr::arrange(linechart1, desc(year), desc(value))
    countries <- linechart1[1:6, 3]
    countries <- countries$country_code
    linechart1 <- dplyr::filter(linechart1, country_code %in% countries)
    linechart1$year <- lubridate::ymd(linechart1$year, truncated = 2L)
    if(title == TRUE){
      ggplot2::ggplot(data = linechart1, ggplot2::aes(x = year, y = value, color = country_name, shape = country_name)) +
        ggplot2::geom_line() +
        ggplot2::geom_point(size = 2, stroke = 1) +
        ggplot2::ylab("")  +
        ggplot2:: xlab("") +
        ggplot2::ggtitle(paste0(unique(linechart1$indicator_name), ": \nEvolution of the ", max(SPI_data$year), " most represented countries")) +
        ggplot2::theme_minimal() +
        ggplot2::guides(fill=FALSE) +
        ggsci::scale_color_uchicago() +
        ggplot2::theme(legend.position="right", plot.title = ggplot2::element_text(size=12))  +
        ggplot2::labs(shape = "Countries", color = "Countries", caption="Source: Warin (2020) & Social Progress Index.")
    } else {
      ggplot2::ggplot(data = linechart1, ggplot2::aes(x = year, y = value, color = country_name, shape = country_name)) +
        ggplot2::geom_line() +
        ggplot2::ylab("")  +
        ggplot2::xlab("") +
        ggplot2::ggtitle("") +
        ggplot2::theme_minimal() +
        ggplot2::guides(fill=FALSE) +
        ggplot2::geom_point(size = 2, stroke = 1) +
        ggsci::scale_color_uchicago() +
        ggplot2::theme(legend.position="right", plot.title = ggplot2::element_text(size=12))  +
        ggplot2::labs(shape = "Countries", color = "Countries", caption="Source: Warin (2020) & Social Progress Index.")
    }
  } else{
    stop("invalid arguments")
  }
}

